File: 0017LetterCombinationsofaPhoneNumber/LetterCombinationsofaPhoneNumber.py
# -*- coding: utf-8 -*-

from typing import List

class LetterCombinationsofaPhoneNumber:
    __KEYS = [
        '', '', "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    ]
    def letterCombinations1(self, digits: "str") -> "List[str]":
        pass

    def letterCombinations3(self, digits: "str") -> "List[str]":
        if len(digits) == 0: return []
        ans = []
        self.__combHelper(digits, ans, '', 0)
        return ans

    def __combHelper(
        self, digits: "str", ans: "List[str]", pre: "str", len_: "int"
    ) -> "None":
        if len_ == len(digits):
            ans.append(pre)#.copy())
            return
        letters = self.__KEYS[int(digits[len_])]
        for c in letters:
            self.__combHelper(digits, ans, pre + c, len_ + 1)


def main():
    test = LetterCombinationsofaPhoneNumber()
    print(test.letterCombinations3("25"))

if __name__ == "__main__":
    main()

File: 0017LetterCombinationsofaPhoneNumber/LetterCombinationsofaPhoneNumber.cpp


#include <vector>
#include <string>

class LetterCombinationsofaPhoneNumber {
public:
    std::vector<std::string> letterCombinations(std::string digits) {
        std::vector<std::string> ans;
        if (digits.size() == 0) return ans;
        combHelper(digits, ans, "", 0);
        return ans;
    }

private:
    void combHelper(
        std::string& digits,
        std::vector<std::string>& ans,
        std::string pre,
        size_t len
    ) {
        if (digits.length() == len) {
            ans.push_back(pre);
            return;
        }
        std::string KEYS[10] = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };
        std::string letters = KEYS[digits[len] - '0'];
        for (size_t i = 0; i < letters.length(); ++i) {
            combHelper(
                digits, ans, pre.append(std::string(1, letters[i])), len + 1
            );
            pre.pop_back();
        }
    }
};
